Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -42,202 +46,460 @@ class SPDHG(Algorithm):
-        tau : positive float, optional, default=None
-            Step size parameter for Primal problem
-        sigma : list of positive float, optional, default=None
-            List of Step size parameters for Dual problem
-        initial : DataContainer, optional, default=None
-            Initial point for the SPDHG algorithm
-        prob : list of floats, optional, default=None
-            List of probabilities. If None each subset will have probability = 1/number of subsets
-        gamma : float
-            parameter controlling the trade-off between the primal and dual step sizes
-
-        **kwargs:
-        norms : list of floats
-            precalculated list of norms of the operators
         '''
         log.info("%s setting up", self.__class__.__name__)
+
         # algorithmic parameters
         self.f = f
         self.g = g
         self.operator = operator
-        self.tau = tau
-        self.sigma = sigma
-        self.prob = prob
-        self.ndual_subsets = len(self.operator)
-        self.gamma = gamma
-        self.rho = .99
-
-        if self.prob is None:
-            self.prob = [1/self.ndual_subsets] * self.ndual_subsets
-
-
-        if self.sigma is None:
-            if norms is None:
-                # Compute norm of each sub-operator
-                norms = [operator.get_item(i,0).norm() for i in range(self.ndual_subsets)]
-            self.norms = norms
-            self.sigma = [self.gamma * self.rho / ni for ni in norms]
-        if self.tau is None:
-            self.tau = min( [ pi / ( si * ni**2 ) for pi, ni, si in zip(self.prob, norms, self.sigma)] )
-            self.tau *= (self.rho / self.gamma)
+
+        if not isinstance(operator, BlockOperator):
+            raise TypeError("operator should be a BlockOperator")
+
+        self._ndual_subsets = len(self.operator)
+        
+
+        # Set up sampler and prob weights from deprecated "prob" argument
+        sampler = self._deprecated_set_prob(deprecated_kwargs, prob_weights, sampler) 
+        
+        
+        self._prob_weights = getattr(sampler, 'prob_weights', prob_weights) 
+        if self._prob_weights is None: 
+            self._prob_weights = [1/self._ndual_subsets]*self._ndual_subsets
+        
+        if  prob_weights is not None and self._prob_weights != prob_weights:
+                    raise ValueError(' You passed a `prob_weights` argument and a sampler with a different attribute `prob_weights`, please remove the `prob_weights` argument.')
+
+        if sampler is None:
+            self._sampler = Sampler.random_with_replacement(
+                len(operator), prob=self._prob_weights)
+        else:
+            self._sampler = sampler
+
+        #Set the norms of the operators
+        self._deprecated_set_norms(deprecated_kwargs) 
+        self._norms = operator.get_norms_as_list()
+        #Check for other kwargs
+        self._deprecated_else(deprecated_kwargs)
```

Comment:
I'm not sure this is worth having as a function, just a check on unused kwargs would maybe suffice if it's needed

Reply:
done